src/v/archival/ntp_archiver_service.cc

```diff
@@ -962,20 +959,32 @@ ss::future<cloud_storage::upload_result> ntp_archiver::do_upload_segment(
         }
     };
 
-    co_return co_await _remote.upload_segment(
-      get_bucket_name(),
-      path,
-      candidate.content_length,
-      std::move(reset_func),
-      fib,
-      lazy_abort_source,
-      _segment_tags);
-}
-
-ss::future<> ntp_archiver::stream_reference::maybe_close_ref() {
-    if (stream_ref.has_value()) {
-        co_await stream_ref->get().close();
+    auto result
+      = co_await _remote
+          .upload_segment(
+            get_bucket_name(),
+            path,
+            candidate.content_length,
+            std::move(reset_func),
+            fib,
+            lazy_abort_source,
+            _segment_tags)
+          .handle_exception_type([](const ss::gate_closed_exception&) mutable {
+              return cloud_storage::upload_result::cancelled;
+          })
+          .handle_exception_type(
+            [](const ss::abort_requested_exception&) mutable {
+                return cloud_storage::upload_result::cancelled;
+            })
+          .handle_exception_type([this, path](const std::exception& e) mutable {
+              vlog(_rtclog.error, "failed to upload segment {}: {}", path, e);
+              return cloud_storage::upload_result::failed;
+          });
+    // Cleanup the stream state if upload_segment didn't use it.
+    if (stream_state.has_value()) {
+        co_await stream_state->close();
```

Comment:
nit: this could be moved to `upload_segment().finally` and then just return the result of the upload inline?

Reply:
Done.